File: ios/Stax/Components/PremiumGate.swift
import SwiftUI

// MARK: – Environment action for triggering paywall

struct ShowPaywallAction {
    let action: () -> Void
    func callAsFunction() { action() }
}

private struct ShowPaywallKey: EnvironmentKey {
    static let defaultValue = ShowPaywallAction(action: {})
}

extension EnvironmentValues {
    var showPaywall: ShowPaywallAction {
        get { self[ShowPaywallKey.self] }
        set { self[ShowPaywallKey.self] = newValue }
    }
}

// MARK: – ViewModifier

struct PremiumGate: ViewModifier {
    @EnvironmentObject var entitlementManager: EntitlementManager

    let feature: Feature
    var sessionPhotoCount: Int?
    var totalSessions: Int?
    var favoritesCount: Int?
    var casinoIndex: Int?
    var onShowPaywall: () -> Void

    func body(content: Content) -> some View {
        let result = entitlementManager.checkLimit(
            for: feature,
            sessionPhotoCount: sessionPhotoCount,
            totalSessions: totalSessions,
            favoritesCount: favoritesCount,
            casinoIndex: casinoIndex
        )

        switch result {
        case .allowed:
            content

        case .softCap(let msg):
            VStack(spacing: 8) {
                content
                UpgradeBanner(message: msg, onUpgrade: onShowPaywall)
            }

        case .blocked:
            UpgradeBanner(
                message: blockedMessage(for: feature),
                onUpgrade: onShowPaywall
            )
        }
    }

    private func blockedMessage(for feature: Feature) -> String {
        switch feature {
        case .scan:          return "You've used all \(FreeTierLimits.maxScansPerDay) free scans today. Upgrade for unlimited."
        case .aiScan:        return "AI Stack Counter is a Premium feature."
        case .sessionCreate: return "You've reached the \(FreeTierLimits.maxSessions)-session free limit."
        case .photoAdd:      return "You've reached the \(FreeTierLimits.maxPhotosPerSession)-photo limit per session."
        case .chipConfig:    return "Upgrade to configure all casinos."
        case .favorites:     return "Upgrade to Premium for unlimited favorites."
        case .shareClean:    return "Clean export is a Premium feature."
        }
    }
}

// MARK: – View extension

extension View {
    func premiumGate(
        feature: Feature,
        sessionPhotoCount: Int? = nil,
        totalSessions: Int? = nil,
        favoritesCount: Int? = nil,
        casinoIndex: Int? = nil,
        onShowPaywall: @escaping () -> Void
    ) -> some View {
        self.modifier(PremiumGate(
            feature: feature,
            sessionPhotoCount: sessionPhotoCount,
            totalSessions: totalSessions,
            favoritesCount: favoritesCount,
            casinoIndex: casinoIndex,
            onShowPaywall: onShowPaywall
        ))
    }
}

File: android/app/src/main/java/com/bitcraftapps/stax/ui/composables/PremiumGate.kt
package com.bitcraftapps.stax.ui.composables

import androidx.compose.foundation.layout.Column
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import com.bitcraftapps.stax.data.billing.Feature
import com.bitcraftapps.stax.data.billing.LimitResult
import com.bitcraftapps.stax.data.billing.LocalEntitlementManager

@Composable
fun PremiumGate(
    feature: Feature,
    onShowPaywall: () -> Unit,
    extraParams: Map<String, Any> = emptyMap(),
    content: @Composable () -> Unit
) {
    val entitlementManager = LocalEntitlementManager.current
    val isPremium by entitlementManager.isPremium.collectAsState()

    val sessionPhotoCount = (extraParams["sessionPhotoCount"] as? Int)
    val totalSessions = (extraParams["totalSessions"] as? Int)
    val favoritesCount = (extraParams["favoritesCount"] as? Int)
    val casinoIndex = (extraParams["casinoIndex"] as? Int)

    val limitResult = entitlementManager.checkLimit(
        feature = feature,
        sessionPhotoCount = sessionPhotoCount,
        totalSessions = totalSessions,
        favoritesCount = favoritesCount,
        casinoIndex = casinoIndex
    )

    when (limitResult) {
        is LimitResult.Allowed -> content()
        is LimitResult.SoftCap -> {
            Column {
                content()
                UpgradeBanner(
                    message = limitResult.message,
                    onUpgrade = onShowPaywall
                )
            }
        }
        is LimitResult.Blocked -> {
            UpgradeBanner(
                message = featureLockMessage(feature),
                onUpgrade = onShowPaywall
            )
        }
    }
}

private fun featureLockMessage(feature: Feature): String = when (feature) {
    Feature.SCAN -> "You've used all 5 free scans today. Upgrade for unlimited."
    Feature.AI_SCAN -> "AI Stack Counter is a Premium feature. Upgrade to unlock."
    Feature.SESSION_CREATE -> "You've reached the 3-session limit. Upgrade for unlimited."
    Feature.PHOTO_ADD -> "You've reached 10 photos for this session. Upgrade for unlimited."
    Feature.CHIP_CONFIG -> "Upgrade to configure all casinos."
    Feature.FAVORITES -> "You've reached 3 favorites. Upgrade for unlimited."
    Feature.SHARE_CLEAN -> "Clean export is a Premium feature. Upgrade to unlock."
}
